### PR TITLE
Add 'no-unused' rule

### PR DESCRIPTION
**Pull request by andy-hanson**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [X] New feature, bugfix, or enhancement
  - [X] Includes tests
- [X] Documentation update

#### Overview of change:

Adds the `no-unused` rule, which detects a variety of unused capabilities; for example, recommends to convert from `public x: string[]` to `private readonly x: ReadonlyArray<string>` if possible. See the description in `noUnusedRule.ts` for more details.

#### Is there anything you'd like reviewers to focus on?

This rule should probably have options to disable certain checks -- I'm thinking there should be flags for:
* Prefer ReadonlyArray/ReadonlySet/ReadonlyMap
* Prefer `readonly` properties
* Prefer `private` properties

Also, the code of this rule could be used for the `no-unused-variable`, `prefer-readonly`, aand `prefer-const` -- and those don't need to be enabled if this rule is.

#### CHANGELOG.md entry:

[new-rule] `no-unused`

### CODE REVIEW

File: test/rules/no-unnecessary-qualifier/test-global.ts.lint [hunk 1]
```diff
@@ -1,6 +1,6 @@
-namespace N {
+namespace Global {
```
Comment (andy-hanson) [inline test/rules/no-unnecessary-qualifier/test-global.ts.lint]: Due to changes to the test runner this is now compiled along with the other files in this directory -- which already declares `namespace N`.

File: src/test.ts [hunk 2]
```diff
@@ -25,49 +25,49 @@ import * as ts from "typescript";
 
 import { Replacement } from "./language/rule/rule";
 import * as Linter from "./linter";
+import { createMultiMap } from "./rules/no-unused/utils";
```
Comment (andy-hanson) [inline src/test.ts]: Needed to change the test runner to be able to test files that import each other -- these must end in `.ts` and not `.lint` or TypeScript can't resolve the modules.

**Comment by giladgray** [comment]
@andy-hanson how is this different from tsconfig `--noUnusedLocals` and `--noUnusedParameters`?

**Comment by giladgray** [comment]
> Also, the code of this rule could be used for the `no-unused-variable`, `prefer-readonly`, aand `prefer-const` -- and those don't need to be enabled if this rule is.

it seems like we already have rules for a lot of the checks here and there's some overlap with TS compiler flags, so i'm wondering why this needs to be a rule at all?

**Comment by aboyton** [comment]
I hit an issue yesterday where doing `foo.push` was marking `foo` as used. If this rule could help with that would be great. https://github.com/Microsoft/TypeScript/issues/24309

**Comment by giladgray** [comment]
@andy-hanson this feels like too much overlap with `tsc`'s existing unused detection (which also got waaay better in 2.9). this also _is_ way too much code for me to feel comfortable maintaining.
 
do we still need this? will close if no answer in two weeks.

**Comment by andy-hanson** [comment]
> tsc's existing unused detection (which also got waaay better in 2.9)

There haven't been any big changes lately in regards to what's detected. I've done most of the unused-related work lately on TS and it has mostly been about improved diagnostic messages and codefixes, but not changing the kinds of things that are errors. (You might be thinking of Microsoft/TypeScript#22204 which made unused errors appear in the editor *without* `--noUnusedLocals` or `--noUnusedParameters` set, but didn't change the behavior when they are set.)

TypeScript is unlikely to ever detect significantly more cases than it does today because it's a compiler rather than a linter. It probably should never have had unused checking in the first place, since this is naturally a linting feature -- the code isn't going to error, but could be cleaned up before release.

TypeScript only detects unused local variables and private properties. (The tslint rules `prefer-const` and `prefer-readonly` similarly don't work for public/exported symbols. `no-unused-variable` just wraps TypeScript's version and has the same problem.) This PR detects *many* more cases because it analyzes an entire program.

TypeScript's unused checking works by adding an `isReferenced` flag to every symbol. If that symbol is touched even once, it is considered used, regardless of *how* it is used. (There is an exception for local variables that are only assigned to, but that doesn't detect the `push`-only case case.) It won't warn on an array that is only ever pushed to, or a mutable array that could be Readonly. It won't warn for properties that could be private or readonly. It won't detect optional properties that are read but never assigned to. It won't detect enum members that are tested for but never provided as a value. It also *never* checks public properties or exports at all.

The PR is large mostly because of the analysis functionality (from the `no-unused` directory.) Perhaps that could be moved to a publicly exported function in the TypeScript codebase that it can call (note this does not mean making these compile errors).

**Comment by JoshuaKGoldberg** [comment]
> which detects a variety of unused capabilities

This PR is a huge step forward _(and honestly technically very impressive)_, but seems like multiple rules smooshed into one. We've had better success with adding multiple single-purpose rules than with larger multi-purpose rules. I can see this rule being split into:

* `no-unused-locals`: new rule comparable to TypeScript's `--noUnusedLocals`
* `no-unused-parameters`: new rule comparable to TypeScript's `--noUnusedParameters`
* `prefer-const`: adding options to improve that rule
* `prefer-private`: new rule
* `prefer-readonly-locals`: new rule
* `prefer-readonly-parameters`: new rule 
* `prefer-readonly-members`: renamed version of the existing `prefer-readonly`, with added options

Given #4534, it feels like this discussion would be more productive if moved into the [typescript-eslint repository](https://typescript-eslint.io). Thoughts?

**Comment by JoshuaKGoldberg** [comment]
Closing for age. Ah well!

**Pull request closed without merging**